plugins/adrv9002.c

```diff
@@ -0,0 +1,1241 @@
+	/* misc */
+	gboolean plugin_detached;
+	gint this_page;
+	gint refresh_timeout;
+	char last_profile[PATH_MAX];
+	struct adrv9002_gtklable temperature;
+	/* rx */
+	struct adrv9002_rx rx_widgets[ADRV002_NUM_CHANNELS];
+	/* tx */
+	struct adrv9002_common tx_widgets[ADRV002_NUM_CHANNELS];
+	/* dac */
+	struct adrv9002_dac_mgmt dac_manager[NUM_MAX_DDS];
+	int n_dacs;
+	/* adc */
+	const char *adc_name[NUM_MAX_ADC];
+	int n_adcs;
+};
+
+#define dialog_box_message(widget, title, msg) { 					\
+	GtkWidget *toplevel = gtk_widget_get_toplevel(widget);				\
+											\
+	if (gtk_widget_is_toplevel(toplevel)) {						\
+		GtkWidget *dialog;							\
+											\
+		dialog = gtk_message_dialog_new(GTK_WINDOW(toplevel),			\
+					GTK_DIALOG_DESTROY_WITH_PARENT,			\
+					GTK_MESSAGE_ERROR, GTK_BUTTONS_CLOSE,		\
+					msg);						\
+											\
+		gtk_window_set_title(GTK_WINDOW(dialog), title);			\
+		gtk_dialog_run(GTK_DIALOG(dialog));					\
+		gtk_widget_destroy (dialog);						\
+	} else {									\
+		printf("Cannot display dialog: Toplevel wigdet not found\n");		\
+	}										\
+}
+
+static void save_widget_value(GtkWidget *widget, struct iio_widget *iio_w)
+{
+	iio_w->save(iio_w);
+	iio_w->update(iio_w);
+}
+
+static void combo_box_manual_update(struct adrv9002_combo_box *combo)
+{
+	char text[512], *item;
+	int ret, i = 0;
+	struct iio_widget *w = &combo->w;
+	GtkWidget *widget = w->widget;
+	gint idx = gtk_combo_box_get_active(GTK_COMBO_BOX(widget));
+	GtkTreeIter iter;
+	GtkTreeModel *model = gtk_combo_box_get_model(GTK_COMBO_BOX(widget));
+	gboolean has_iter;
+
+	ret = iio_channel_attr_read(w->chn, w->attr_name,
+				    text, sizeof(text));
+	if (ret < 0)
+		return;
+
+	has_iter = gtk_tree_model_get_iter_first(model, &iter);
+	while (has_iter) {
+		gtk_tree_model_get(model, &iter, 0, &item, -1);
+		if (strcmp(text, item) == 0) {
+			if (i != idx) {
+				/*
+				 * This assumes the gtk signal is connected prior to call this function.
+				 * If the signal is connected after this API is called and we get an update,
+				 * we won't save the attribute value, the next time the user changes it in
+				 * the GUI.
+				 */
+				combo->m_update = true;
+				gtk_combo_box_set_active(GTK_COMBO_BOX(widget), i);
+			}
+
+			g_free(item);
+			break;
+		}
+		g_free(item);
+		i++;
+		has_iter = gtk_tree_model_iter_next(model, &iter);
+	}
+}
+
+/*
+ * This checks if the there was any manual or automatic update before
+ * writing the value in the device. Furthermore, it checks if were successful
+ * in writing the value.
+ */
+static void combo_box_save(GtkWidget *widget, struct adrv9002_combo_box *combo)
+{
+	if (combo->m_update) {
+		combo->m_update = false;
+		return;
+	}
+	combo->w.save(&combo->w);
+	combo_box_manual_update(combo);
+}
+
+static void save_gain_ctl(GtkWidget *widget, struct adrv9002_common *chann)
+{
+	char *gain_ctl;
+
+	combo_box_save(widget, &chann->gain_ctrl);
+
+	gain_ctl = gtk_combo_box_text_get_active_text(
```

Comment:
gain_ctl needs to be freed

Reply:
Ohh then I need to fix this in some other places. I just copied it from another plugin where I guess we have the same problem.

Reply:
I'll write this down to make sure we will go over all plugins and free where it is needed.

Reply:
ack. In adrv9009 we definitely have the same issue